### ISSUE TITLE
Improve battery consumption of live wallpaper

### ISSUE THREAD

**Issue by lone-cloud** [breezy-weather/breezy-weather#1006]
### Steps to reproduce

Get latest breezy weather (currently 5.2.1).
Turn off battery optimizations.
Add live wallpaper.
Add a couple of widgets.
3 hour weather refresh rate.

### Expected behavior

Battery usage should be minimal.

### Actual behavior

Breezy weather is constantly the top battery consumer on my phone. I don't see how that can be possible for a simple weather app.
![image](https://github.com/breezy-weather/breezy-weather/assets/595980/989c236c-a13b-430e-9483-e0add864c648)

I'm not running into errors and everything seems to be working great. If the live wallpaper is such a crazy battery hog, it should be added to the FAQ. I was not expecting for it to be anywhere near the top in terms of battery consumption. Note that I've also turned off battery optimizations for this app as was suggested by the app itself. Was that a mistake?

### Weather source used

AccuWeather

### Breezy Weather version

5.2.1

### Android version

latest Android 14 (GOS)

### Device

Pixel 8

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open OR closed issue.
- [X] I checked that my issue is not covered in the [Frequently Asked Questions](https://github.com/breezy-weather/breezy-weather/blob/main/HELP.md) document.
- [X] I have updated the app to **[version 5.2.1 or later](https://github.com/breezy-weather/breezy-weather/releases/latest)** and I can reproduce the issue in this version.
- [X] I wrote my request in English, and any screenshot I attached are also in English (language can be changed from Settings > Appearance), unless my issue only happens with a specific language. I can write in my native language under the English text, if necessary.
- [X] I understand that this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY, so I may not get a positive reply to my request, if any at all.

**Comment by papjul** [comment]
I don't use the live wallpaper but given how it was written, I wouldn't be surprised if it was battery-consuming. Didn't you have the same issue in Geometric Weather?

Please try the following:
1) Not use live wallpaper
2) Set manual weather and day/nighttime for the wallpaper

Please report if the battery is better in cases 1 and 2.

**Comment by papjul** [comment]
See my concerns here when I found this piece of code:
https://github.com/breezy-weather/breezy-weather/blob/v5.2.1/app/src/main/java/org/breezyweather/wallpaper/MaterialLiveWallpaperService.kt#L358-L359

I also notice that you have a phone with 120 Hz refresh rate. I will revert the change I made to use native refresh rate to cap that to 60 Hz instead, which should at least make it half less battery-consuming.

**Comment by papjul** [comment]
> I also notice that you have a phone with 120 Hz refresh rate. I will revert the change I made to use native refresh rate to cap that to 60 Hz instead, which should at least make it half less battery-consuming.

Implemented in eaadfed72cabf41de9e5e0be1be48d7acb43be7c

**Comment by lone-cloud** [comment]
I've never used geometric weather, so I can't comment on that.
I am indeed using a 120hz device and I do have "smooth display" enabled.
Thanks for the new release. I'll switch over to 5.2.2 and see how the battery life is affected before considering getting rid of the live wallpaper.

**Comment by lone-cloud** [comment]
OK, I've verified that it was definitely the live wallpaper that was a massive battery hog for me. Without it, breezy weather is taking up less than 1% of the battery on my phone, which is what I would expect.
Just a thought, but I live in a pretty rainy place (Vancouver) and the live wallpaper rain animation is quite intense. I would think that it would consume more battery life than the other weather types as there's very little animation happening for the sunny weather, as an example. I would think that android is smart enough to not animate something at 60/120hz if the picture is largely static like the sunny weather live wallpaper.
Anyways, I'm switching to a static image as I want my battery to last. Thanks for the help.

**Comment by papjul** [comment]
The rain animation and sunny animation are rather similar in terms of elements. Even if the sun is small to your eyes, it's still running about 5 layers of star-shaped suns if I remember correctly. Additionally, what consumes the most battery (I think) is that it queries the database too often to get updates of the current weather and show the correct weather. It would be worth trying to set a manual wallpaper (both weather condition and time of the day) to see how better it performs.

**Comment by StevenAlexander44** [comment]
![Screenshot_20240621-144322_Settings](https://github.com/breezy-weather/breezy-weather/assets/64290624/0e9bafc3-1310-4f99-aef1-90375ff6a1d0)
Pixel 4 XL running LineageOS 21 (Android 14) on a sunny day with live wallpaper on, but my screentime looking at the wallpaper is only about 5 minutes.

**Comment by papjul** [comment]
Yes, we know. 5 minutes of wallpaper time is `60 * 60 * 5 = 18 000`. 18 000 queries of the database and redraws is enormous. Tl;dr: don't use the live wallpaper.

**Comment by owosoup** [comment]
I was wondering, could an option to lower the refresh rate to 30hz be available? Strangely live wallpapers don't drain battery that much on my device, but it takes a bit of GPU power and slows down kvaesitso. I'm on a realme 9 with realme UI (android 13).

**Comment by papjul** [comment]
The way it is designed, the database is in "pull" mode while it should be in "push" mode instead (push new data to the live wallpaper when data is refreshed instead of making the live wallpaper check the database 60 times per second).
Also, I'm pretty sure there is room for improvement of how the canvas is drawn, as too much seems to happen in the refresh function, because the issue still happens when not querying the database (manual mode).

That's what needs to be done. Controlling the rate has no guarantee to make any improvement, because that's not the reason the issue is happening 

I thought I had already done it before but I'm opening this to contributions, as I am not very familiar with this part of the code and I'm not using the wallpaper, so some people might take a better look at it ;)

**Comment by papjul** [comment]
The live wallpaper consumes way too much memory/battery, even leading to crashes in some cases.
For this reason, animations will be removed from the live wallpaper in next version (the wallpaper will still be available, but in static way).
See #1665 if you have some expertise with shaders for a potential more efficient way to bring them back.
